### PR TITLE
AWS S3: versioning support

### PR DESCRIPTION
**Pull request by sfali**
Support for download, upload, copy, and delete for a specific version, when the bucket is enabled to support versioning.

#944

### CODE REVIEW

File: docs/src/main/paradox/s3.md [hunk 1]
```diff
@@ -93,9 +101,15 @@ Copy an S3 object from source bucket to target bucket using multi part copy uplo
 Scala
 : @@snip ($alpakka$/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala) { #multipart-copy }
 
+Scala
+: @@snip ($alpakka$/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala) { #multipart-copy-with-source-version }
+
```
Comment (ennru) [inline docs/src/main/paradox/s3.md]: You should make this two snippet sections, and add some text explaining what it does.

File: s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala [hunk 2]
```diff
@@ -195,10 +223,14 @@ final class S3Client(s3Settings: S3Settings, system: ActorSystem, mat: Materiali
    */
   def request(bucket: String,
               key: String,
+              versionId: String,
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala]: Wouldn't it be better to go with Optional instead of a nullable String?
We might deprecate the `request` without the parameter.

File: s3/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+version: '3.1'
+
+services:
+  minio1:
```
Comment (ennru) [inline s3/docker-compose.yaml]: Did you implement integration tests against these minio servers?
I so, please send a separate PR with those and remove this section for now.

Comment (sfali) [reply s3/docker-compose.yaml]: Yes I ran against integration tests these minio servers. I noticed that you have an issue, #938, open to run integration tests in CI, in that case there should be a way to create buckets on startup and delete once done.

I'll remove and will send new PR for this.

**Comment by sfali** [comment]
As for creating a PR for docker compose file, as I mentioned in my comments if you are planning to run in CI/CD pipeline then functions need to be added to create and delete buckets at the start and end respectively, let me know about your thoughts.

**Comment by ennru** [comment]
Thank you for growing the S3 connector even more.
